File: problem_solving/sliding_window/py/1493_longest_subarray_after_deleting_1.py
'''
https://leetcode.com/problems/longest-subarray-of-1s-after-deleting-one-element/

sliding window
r- for loop
'''


def longest_subarray_after_deleting_ones(nums: list[int]) -> int:
    N=len(nums)
    l = 0
    zeros = 0
    max_len = 0
    for r in range(N):
        if nums[r] == 0:
            zeros += 1
            while l <= N and zeros > 1:
                if nums[l] == 0:
                    zeros -= 1
                l += 1
        max_len = max(max_len, r-l)
    return max_len    

nums = [1,1,0,1]
#Output: 3
result = longest_subarray_after_deleting_ones(nums)
print("result= " + str(result))

nums = [0,1,1,1,0,1,1,0,1]
#Output: 5
result = longest_subarray_after_deleting_ones(nums)
print("result= " + str(result))


nums = [1,1,1]
#Output: 2
result = longest_subarray_after_deleting_ones(nums)
print("result= " + str(result))

File: problem_solving/sliding_window/cpp/1493_longest_subarray_after_deleting_1.cpp
/*'''
https://leetcode.com/problems/longest-subarray-of-1s-after-deleting-one-element/

sliding window
r- for loop
'''
*/

#include "../../../cpp/common_header.h"
#include "../../../cpp/utils/utils.h"


int longest_subarray_after_deleting_ones(std::vector<int>& nums) {
    int len = nums.size();
    int l = 0;
    int longest = 0;
    int zeros = 0;
    for(int r = 0; r < len; ++r) {
        if(nums[r] == 0) {
            zeros++;
        }
        while(zeros > 1) {
            if(nums[l] == 0) {
                zeros--;
            }
            l++;
        }
        longest = std::max(longest, r-l);//r-l+1 -1 => since we need to delete 1 
    }
    return longest;
}

int main() {
    std::vector<int> nums = {1,1,0,1};
    //#Output: 3
    int result = longest_subarray_after_deleting_ones(nums);
    std::cout << "result= " << result << std::endl;
    
    nums = {0,1,1,1,0,1,1,0,1};
    //#Output: 5
    result = longest_subarray_after_deleting_ones(nums);
    std::cout << "result= " << result << std::endl;
    
    
    nums = {1,1,1};
    //#Output: 2
    result = longest_subarray_after_deleting_ones(nums);
    std::cout << "result= " << result << std::endl;
}
